File: agent/lib/stack.ts
import * as fs from "fs";
import * as path from "path";
import { sanitizeForDocker } from "./docker.js";
import { runCommand, formatOutput } from "./run-command.js";

function envToYaml(env: Record<string, string> | undefined): string {
  if (!env || Object.keys(env).length === 0) return "";
  const lines = Object.entries(env)
    .filter(([, v]) => v !== undefined && v !== null)
    .map(([k, v]) => `      ${k}: ${typeof v === "string" && (v.includes("\n") || v.includes(":") || v.includes("#") ? JSON.stringify(String(v)) : String(v))}`);
  if (lines.length === 0) return "";
  return "    environment:\n" + lines.join("\n") + "\n";
}

function resourcesBlock(cpuLimit?: string | null, memoryLimit?: string | null): string {
  if (!cpuLimit?.trim() && !memoryLimit?.trim()) return "";
  const c = cpuLimit?.trim();
  const m = memoryLimit?.trim();
  const parts: string[] = [];
  if (c) parts.push(`        cpus: "${c}"`);
  if (m) parts.push(`        memory: ${m}`);
  if (parts.length === 0) return "";
  return "      resources:\n        limits:\n" + parts.join("\n") + "\n";
}

export function generateStackYaml(
  stackName: string,
  imageTag: string,
  port: number = 80,
  options?: { env?: Record<string, string>; replicas?: number; domain?: string | null; cpuLimit?: string | null; memoryLimit?: string | null }
): string {
  const safe = sanitizeForDocker(stackName);
  const domain = options?.domain?.trim();
  const host = domain ? `${safe}.${domain}` : `${safe}.localhost`;
  const envBlock = envToYaml(options?.env);
  const replicas = Math.min(32, Math.max(1, options?.replicas ?? 1));
  const resources = resourcesBlock(options?.cpuLimit, options?.memoryLimit);
  // In Docker Swarm, service name seen by Traefik is stackname_servicename (e.g. vite_app)
  const traefikServiceName = `${safe}_app`;
  // HTTP only here; SSL/TLS is handled by Traefik (gateway) via app/traefik and app API traefik config
  return `version: "3.9"

networks:
  ${NETWORK}:
    external: true

services:
  app:
    image: ${imageTag}
${envBlock}    networks:
      - ${NETWORK}
    deploy:
      replicas: ${replicas}
${resources}      restart_policy:
        condition: on-failure
      labels:
        - "traefik.enable=true"
        - "traefik.swarm.network=${NETWORK}"
        - "traefik.http.routers.${safe}.rule=Host(\`${host}\`)"
        - "traefik.http.routers.${safe}.entrypoints=web"
        - "traefik.http.routers.${safe}.service=${traefikServiceName}"
        - "traefik.http.services.${traefikServiceName}.loadbalancer.server.port=${port}"
`;
}

const NETWORK = process.env.TRAEFIK_NETWORK ?? "web";

/** Create the overlay network if it does not exist (required for stack deploy). Uses --attachable so diagnostics can run curl from a temp container. */
function ensureSwarmNetwork(): void {
  const result = runCommand(`docker network create --driver overlay --attachable ${NETWORK}`);
  if (!result.success && !result.stderr.includes("already exists")) {
    console.warn("[stack] Could not create network:", result.stderr || result.stdout);
  }
}

export function deployStack(stackName: string, yamlContent: string): void {
  ensureSwarmNetwork();
  const dir = path.join(process.cwd(), ".stack-tmp");
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  const safe = sanitizeForDocker(stackName);
  const file = path.join(dir, `${safe}-stack.yml`);
  fs.writeFileSync(file, yamlContent);
  try {
    const result = runCommand(`docker stack deploy -c "${file}" ${safe}`);
    if (!result.success) throw new Error(`docker stack deploy failed:\n${formatOutput(result)}`);
  } finally {
    try {
      fs.unlinkSync(file);
    } catch {}
  }
}

export function sanitizeStackName(name: string): string {
  return sanitizeForDocker(name);
}

const TRAEFIK_STACK_NAME = "traefik";

export function deployTraefikStack(yamlContent: string): void {
  ensureSwarmNetwork();
  const dir = path.join(process.cwd(), ".stack-tmp");
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  const file = path.join(dir, "traefik-stack.yml");
  fs.writeFileSync(file, yamlContent);
  try {
    const result = runCommand(`docker stack deploy -c "${file}" ${TRAEFIK_STACK_NAME}`);
    if (!result.success) throw new Error(`traefik deploy failed:\n${formatOutput(result)}`);
  } finally {
    try {
      fs.unlinkSync(file);
    } catch {}
  }
}

export function removeTraefikStack(): void {
  const result = runCommand(`docker stack rm ${TRAEFIK_STACK_NAME}`);
  if (!result.success && !result.stderr?.includes("Nothing found")) {
    throw new Error(`traefik remove failed:\n${formatOutput(result)}`);
  }
}

export function isTraefikRunning(): boolean {
  const result = runCommand("docker stack ps traefik --no-trunc -q");
  if (!result.success) return false;
  const out = (result.stdout || "").trim();
  return out.length > 0;
}

File: agent-go/internal/stack/stack.go
package stack

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/marcuwynu23/podfire/agent-go/internal/docker"
	"github.com/marcuwynu23/podfire/agent-go/internal/run"
)

var getEnv = func(key string) string { return "" }

// SetGetEnv sets the env getter (from main).
func SetGetEnv(fn func(string) string) {
	getEnv = fn
}

func network() string {
	n := getEnv("TRAEFIK_NETWORK")
	if n != "" {
		return n
	}
	return "web"
}

func envToYaml(env map[string]string) string {
	if len(env) == 0 {
		return ""
	}
	var lines []string
	for k, v := range env {
		if v == "" {
			continue
		}
		// Quote if contains newline, colon, or #
		if strings.ContainsAny(v, "\n:#") {
			v = strconv.Quote(v)
		}
		lines = append(lines, "      "+k+": "+v)
	}
	if len(lines) == 0 {
		return ""
	}
	return "    environment:\n" + strings.Join(lines, "\n") + "\n"
}

func resourcesBlock(cpuLimit, memoryLimit string) string {
	c := strings.TrimSpace(cpuLimit)
	m := strings.TrimSpace(memoryLimit)
	if c == "" && m == "" {
		return ""
	}
	var parts []string
	if c != "" {
		parts = append(parts, `        cpus: "`+c+`"`)
	}
	if m != "" {
		parts = append(parts, "        memory: "+m)
	}
	if len(parts) == 0 {
		return ""
	}
	return "      resources:\n        limits:\n" + strings.Join(parts, "\n") + "\n"
}

// GenerateStackYaml produces the Docker Stack YAML for the app service.
func GenerateStackYaml(stackName, imageTag string, port int, opts *StackOptions) string {
	safe := docker.SanitizeForDocker(stackName)
	domain := ""
	if opts != nil && opts.Domain != "" {
		domain = strings.TrimSpace(opts.Domain)
	}
	host := safe + ".localhost"
	if domain != "" {
		host = safe + "." + domain
	}
	envBlock := ""
	if opts != nil && len(opts.Env) > 0 {
		envBlock = envToYaml(opts.Env)
	}
	replicas := 1
	if opts != nil && opts.Replicas >= 1 && opts.Replicas <= 32 {
		replicas = opts.Replicas
	}
	cpuLimit := ""
	memoryLimit := ""
	if opts != nil {
		cpuLimit = opts.CPULimit
		memoryLimit = opts.MemoryLimit
	}
	resources := resourcesBlock(cpuLimit, memoryLimit)
	traefikServiceName := safe + "_app"
	net := network()
	hostRule := "Host(`" + host + "`)"
	return `version: "3.9"

networks:
  ` + net + `:
    external: true

services:
  app:
    image: ` + imageTag + `
` + envBlock + `    networks:
      - ` + net + `
    deploy:
      replicas: ` + strconv.Itoa(replicas) + `
` + resources + `      restart_policy:
        condition: on-failure
      labels:
        - "traefik.enable=true"
        - "traefik.swarm.network=` + net + `"
        - "traefik.http.routers.` + safe + `.rule=` + hostRule + `"
        - "traefik.http.routers.` + safe + `.entrypoints=web"
        - "traefik.http.routers.` + safe + `.service=` + traefikServiceName + `"
        - "traefik.http.services.` + traefikServiceName + `.loadbalancer.server.port=` + strconv.Itoa(port) + `"
`
}

// StackOptions for GenerateStackYaml.
type StackOptions struct {
	Env         map[string]string
	Replicas    int
	Domain      string
	CPULimit    string
	MemoryLimit string
}

func ensureSwarmNetwork() {
	net := network()
	res := run.Run("docker network create --driver overlay --attachable "+net, "")
	if !res.Success && !strings.Contains(res.Stderr, "already exists") {
		// log warning
	}
}

// DeployStack writes yaml to a temp file and runs docker stack deploy.
func DeployStack(stackName, yamlContent string) error {
	ensureSwarmNetwork()
	dir := filepath.Join(".", ".stack-tmp")
	if err := os.MkdirAll(dir, 0755); err != nil {
		return err
	}
	safe := docker.SanitizeForDocker(stackName)
	file := filepath.Join(dir, safe+"-stack.yml")
	if err := os.WriteFile(file, []byte(yamlContent), 0600); err != nil {
		return err
	}
	defer os.Remove(file)
	res := run.Run(`docker stack deploy -c "`+file+`" `+safe, "")
	if !res.Success {
		return fmt.Errorf("docker stack deploy failed:\n%s", run.FormatOutput(res))
	}
	return nil
}

// SanitizeStackName returns a Docker-safe stack name.
func SanitizeStackName(name string) string {
	return docker.SanitizeForDocker(name)
}

const traefikStackName = "traefik"

// DeployTraefikStack deploys the Traefik stack.
func DeployTraefikStack(yamlContent string) error {
	ensureSwarmNetwork()
	dir := filepath.Join(".", ".stack-tmp")
	if err := os.MkdirAll(dir, 0755); err != nil {
		return err
	}
	file := filepath.Join(dir, "traefik-stack.yml")
	if err := os.WriteFile(file, []byte(yamlContent), 0600); err != nil {
		return err
	}
	defer os.Remove(file)
	res := run.Run("docker stack deploy -c "+file+" "+traefikStackName, "")
	if !res.Success {
		return fmt.Errorf("traefik deploy failed:\n%s", run.FormatOutput(res))
	}
	return nil
}

// RemoveTraefikStack removes the Traefik stack.
func RemoveTraefikStack() error {
	res := run.Run("docker stack rm "+traefikStackName, "")
	if !res.Success && !strings.Contains(res.Stderr, "Nothing found") {
		return fmt.Errorf("traefik remove failed:\n%s", run.FormatOutput(res))
	}
	return nil
}

// IsTraefikRunning returns true if the traefik stack has running tasks.
func IsTraefikRunning() bool {
	res := run.Run("docker stack ps traefik --no-trunc -q", "")
	if !res.Success {
		return false
	}
	return strings.TrimSpace(res.Stdout) != ""
}

// ParseEnvMap converts a JSON object to map[string]string (for stack options).
func ParseEnvMap(data []byte) (map[string]string, error) {
	var m map[string]string
	if err := json.Unmarshal(data, &m); err != nil {
		return nil, err
	}
	return m, nil
}
